File: asm/src/main/java/dev/openrs2/asm/MemberRef.java
package dev.openrs2.asm;

import java.util.Objects;

import org.objectweb.asm.tree.ClassNode;
import org.objectweb.asm.tree.FieldInsnNode;
import org.objectweb.asm.tree.FieldNode;
import org.objectweb.asm.tree.MethodInsnNode;
import org.objectweb.asm.tree.MethodNode;

public final class MemberRef {
	private final String owner, name, desc;

	public MemberRef(ClassNode clazz, FieldNode field) {
		this(clazz.name, field.name, field.desc);
	}

	public MemberRef(ClassNode clazz, MethodNode method) {
		this(clazz.name, method.name, method.desc);
	}

	public MemberRef(FieldInsnNode fieldInsn) {
		this(fieldInsn.owner, fieldInsn.name, fieldInsn.desc);
	}

	public MemberRef(MethodInsnNode methodInsn) {
		this(methodInsn.owner, methodInsn.name, methodInsn.desc);
	}

	public MemberRef(String owner, MemberDesc desc) {
		this(owner, desc.getName(), desc.getDesc());
	}

	public MemberRef(String owner, String name, String desc) {
		this.owner = owner;
		this.name = name;
		this.desc = desc;
	}

	public String getOwner() {
		return owner;
	}

	public String getName() {
		return name;
	}

	public String getDesc() {
		return desc;
	}

	@Override
	public boolean equals(Object o) {
		if (this == o) {
			return true;
		}
		if (o == null || getClass() != o.getClass()) {
			return false;
		}
		var memberRef = (MemberRef) o;
		return owner.equals(memberRef.owner) &&
			name.equals(memberRef.name) &&
			desc.equals(memberRef.desc);
	}

	@Override
	public int hashCode() {
		return Objects.hash(owner, name, desc);
	}

	@Override
	public String toString() {
		return String.format("%s %s.%s", desc, owner, name);
	}
}

File: asm/src/main/java/dev/openrs2/asm/MemberRef.kt
package dev.openrs2.asm

import org.objectweb.asm.tree.ClassNode
import org.objectweb.asm.tree.FieldInsnNode
import org.objectweb.asm.tree.FieldNode
import org.objectweb.asm.tree.MethodInsnNode
import org.objectweb.asm.tree.MethodNode

public data class MemberRef(val owner: String, val name: String, val desc: String) : Comparable<MemberRef> {
    public constructor(clazz: ClassNode, field: FieldNode) : this(clazz.name, field.name, field.desc)
    public constructor(clazz: ClassNode, method: MethodNode) : this(clazz.name, method.name, method.desc)
    public constructor(fieldInsn: FieldInsnNode) : this(fieldInsn.owner, fieldInsn.name, fieldInsn.desc)
    public constructor(methodInsn: MethodInsnNode) : this(methodInsn.owner, methodInsn.name, methodInsn.desc)
    public constructor(owner: String, desc: MemberDesc) : this(owner, desc.name, desc.desc)

    override fun compareTo(other: MemberRef): Int {
        var result = owner.compareTo(other.owner)
        if (result != 0) {
            return result
        }

        result = name.compareTo(other.name)
        if (result != 0) {
            return result
        }

        return desc.compareTo(other.desc)
    }

    override fun toString(): String {
        return "$owner.$name $desc"
    }

    public companion object {
        private val STRING_REGEX = Regex("([^.]+)[.]([^ ]+) (.+)")

        public fun fromString(str: String): MemberRef {
            val result = STRING_REGEX.matchEntire(str)
            require(result != null)
            return MemberRef(result.groupValues[1], result.groupValues[2], result.groupValues[3])
        }
    }
}
